### PR TITLE
UI builder v2

### PR DESCRIPTION
**Pull request by KYPremco**

### CODE REVIEW

File: TimberAPI/DependencySystem/DependencyRegistry.cs [hunk 1]
```diff
@@ -42,20 +37,23 @@ public void AddConfigurator(IConfigurator configurator, SceneEntryPoint entryPoi
...
             InstallAll(containerDefinition, SceneEntryPoint.InGame);
+            InstallAll(containerDefinition, SceneEntryPoint.Global);
         }
 
         [HarmonyPostfix]
         [HarmonyPatch(typeof(MainMenuSceneConfigurator), "Configure")]
         static void InjectIntoMainMenuScene(IContainerDefinition containerDefinition)
         {
             InstallAll(containerDefinition, SceneEntryPoint.MainMenu);
+            InstallAll(containerDefinition, SceneEntryPoint.Global);
```
Comment (hawkfalcon) [inline TimberAPI/DependencySystem/DependencyRegistry.cs:48]: Smart

File: TimberAPI/UIBuilderSystem/ElementSystem/BaseElementBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+using Timberborn.CoreUI;
+using TimberbornAPI.AssetLoaderSystem.AssetSystem;
+using TimberbornAPI.Common;
+using TimberbornAPI.UIBuilderSystem.PresetSystem;
+using UnityEngine.UIElements;
+using LocalizableSlider = TimberbornAPI.UIBuilderSystem.CustomElements.LocalizableSlider;
+
+namespace TimberbornAPI.UIBuilderSystem.ElementSystem
+{
+    public abstract class BaseElementBuilder<TElement, TBuilder> where TElement : VisualElement where TBuilder : BaseElementBuilder<TElement, TBuilder>
```
Comment (hawkfalcon) [inline TimberAPI/UIBuilderSystem/ElementSystem/BaseElementBuilder.cs:11]: This is really cool

File: TimberAPI/UIBuilderSystem/PresetSystem/ButtonPresetFactory.cs [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+                .AddClass(TimberApiStyle.Buttons.Active.SpeedButton3Active)
+                .AddClass(TimberApiStyle.Sounds.Click)
+                .SetName(name)
+                .SetHeight(size == default ? new Length(24, Pixel) : size)
+                .SetWidth(size == default ? new Length(24, Pixel) : size);
+            builder?.Invoke(button);
+            return button.Build();
+        }
+        
+        public LocalizableButton BugTracker(string name = null, Length size = default, Action<ButtonBuilder> builder = default)
```
Comment (hawkfalcon) [inline TimberAPI/UIBuilderSystem/PresetSystem/ButtonPresetFactory.cs:465]: what is this

Comment (KYPremco) [reply TimberAPI/UIBuilderSystem/PresetSystem/ButtonPresetFactory.cs:465]: Its a button, with a bug. the one for reporting bugs.

File: TimberAPI/UIBuilderSystem/PresetSystem/LabelPresetFactory.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+using static UnityEngine.UIElements.Length.Unit;
+
+namespace TimberbornAPI.UIBuilderSystem.PresetSystem
+{
+    public class LabelPresetFactory
+    {
+        private readonly ComponentBuilder _componentBuilder;
+        
+
+        public LabelPresetFactory(ComponentBuilder componentBuilder)
```
Comment (hawkfalcon) [inline TimberAPI/UIBuilderSystem/PresetSystem/LabelPresetFactory.cs:15]: The preset stuff is a great idea.

File: TimberAPI/UIBuilderSystem/UIBuilderConfigurator.cs [hunk 5]
```diff
@@ -1,14 +1,16 @@
...
 {
     public class UIBuilderConfigurator : IConfigurator
     {
         public void Configure(IContainerDefinition containerDefinition)
         {
-            containerDefinition.Bind<IUIBoxBuilder>().To<UIBoxBuilder>().AsTransient();
-            containerDefinition.Bind<IUIBuilder>().To<UIBuilder>().AsSingleton();
-            containerDefinition.Bind<IElementFactory>().To<ElementFactory>().AsSingleton();
+            containerDefinition.Bind<UIBoxBuilder>().To<UIBoxBuilder>().AsTransient();
+            containerDefinition.Bind<UIBuilder>().To<UIBuilder>().AsSingleton();
```
Comment (hawkfalcon) [inline TimberAPI/UIBuilderSystem/UIBuilderConfigurator.cs:11]: is the To necessary

File: TimberAPIExample/Assets/enUS.txt [hunk 6]
```diff
@@ -0,0 +1,35 @@
+ID,Text,Comment
```
Comment (hawkfalcon) [inline TimberAPIExample/Assets/enUS.txt:1]: OOOH a label file in the examples, nice

File: TimberAPIExample/Examples/UIBuilderExample/UIBuilderPreviewPanel/Previews/BackgroundPreview.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        public string GetPreviewName()
+        {
+            return "Backgrounds + scale";
+        }
+
+        public VisualElement GetPreview()
+        {
+            FieldInfo[] backgrounds = typeof(TimberApiStyle.Backgrounds).GetFields(BindingFlags.Static | BindingFlags.Public);
```
Comment (hawkfalcon) [inline TimberAPIExample/Examples/UIBuilderExample/UIBuilderPreviewPanel/Previews/BackgroundPreview.cs:30]: <mind blown emoji>

Comment (hawkfalcon) [inline TimberAPIExample/Examples/UIBuilderExample/UIBuilderPreviewPanel/Previews/BackgroundPreview.cs:30]: <mind blown emoji>

Comment (hawkfalcon) [inline TimberAPIExample/Examples/UIBuilderExample/UIBuilderPreviewPanel/Previews/BackgroundPreview.cs:30]: <mind blown emoji>

File: TimberAPIExample/TimberApiExamplePlugin.cs [hunk 8]
```diff
@@ -34,6 +36,10 @@ public void Awake()
             TimberAPI.DependecyRegistry.AddConfigurator(new ExampleConfigurator());
             // Add a label to localization
             TimberAPI.Localization.AddLabel("ExampleMod.ToolGroups.ExampleToolGroup", "Example Label");
+            
+            // Harmony patches
+            new Harmony("com.timberapi.examples").PatchAll();
```
Comment (hawkfalcon) [inline TimberAPIExample/TimberApiExamplePlugin.cs:41]: Wait did I miss something

**Comment by hawkfalcon** [comment]
Amazing job, can't wait to use it

**Pull request merged**